File: frontend/src/lib/splitting.ts
/**
 * Which steps fan one lane out into several files.
 *
 * Mirrors app/pipeline/splitting.py so the builder can warn *before* a step is
 * added rather than letting the backend reject the next one with no
 * explanation. The backend stays the authority -- this only decides what the
 * dialog says.
 *
 * A fan-out ends the pipeline: the lane is now N files, not one stream, and
 * carrying every part through the remaining steps would multiply the work by a
 * number nobody chose (SplitSeq.group on a free-text annotation is one file
 * per distinct value). The parts are the run's output and each is downloadable.
 */

/** How many outputs a max_count parameter asks for. */
function countOf(value: unknown): number {
  if (value == null || value === "") return 0;
  if (Array.isArray(value)) return value.filter((v) => v != null && v !== "").length;
  if (typeof value === "string") {
    return value.split(",").map((v) => v.trim()).filter(Boolean).length;
  }
  return 1;
}

const isBlank = (v: unknown) => v == null || v === "";

/** True when this step, with these parameters, writes more than one file. */
export function stepSplitsOutput(step: string, params: Record<string, unknown>): boolean {
  switch (step) {
    case "SplitSeq.count":
      return true;
    case "SplitSeq.group":
      // A numeric threshold is the bounded two-way under/at-least split (the
      // DUPCOUNT >= 2 idiom), which carries the at-least part forward.
      return isBlank(params.threshold);
    case "SplitSeq.sort":
      return !isBlank(params.max_count);
    case "SplitSeq.sample":
    case "SplitSeq.samplepair":
      return countOf(params.max_count) > 1;
    default:
      return false;
  }
}

/** One sentence saying why this step will end the pipeline. */
export function splitReason(step: string, params: Record<string, unknown>): string | null {
  if (!stepSplitsOutput(step, params)) return null;
  switch (step) {
    case "SplitSeq.count": {
      const max = params.max_count;
      return isBlank(max)
        ? "This writes one file per partition."
        : `This writes one file per ${max} reads — 20,000 reads at ${max} gives ${Math.ceil(
            20000 / Number(max),
          )} files.`;
    }
    case "SplitSeq.group": {
      const field = params.field ? String(params.field).toUpperCase() : "the field";
      return `Without a threshold this writes one file per distinct value of ${field}.`;
    }
    case "SplitSeq.sort":
      return `Sorting with a max count of ${params.max_count} also partitions the reads into separate files.`;
    default:
      return `Asking for ${countOf(params.max_count)} sample sizes writes one output per size.`;
  }
}

/** The note shown under a step that will end the pipeline. */
export function splitWarning(step: string, params: Record<string, unknown>): string | null {
  const reason = splitReason(step, params);
  if (!reason) return null;
  return `${reason} The pipeline ends here — every part is downloadable, and you can start a new run from whichever one you need.`;
}

/**
 * Whether a step *might* split, from its contract alone.
 *
 * "always" is unconditional; "conditional" depends on the parameters, so the
 * dialog shows a softer note that sharpens once a value is typed.
 */
export type SplitMode = "always" | "conditional" | null;

export function splitHint(mode: SplitMode): string | null {
  if (mode === "always") {
    return "This step splits its output into several files and ends the pipeline.";
  }
  if (mode === "conditional") {
    return "Depending on the parameters below, this step may split its output into several files and end the pipeline.";
  }
  return null;
}

File: backend/app/pipeline/splitting.py
"""
Which steps fan one lane out into several files, and what that means.

pRESTO's SplitSeq subcommands do not all behave alike. ``select`` filters and
writes one file; ``count`` and ``group`` shard the lane into an arbitrary
number of parts (20 000 reads at max_count=4 000 is five files, and a
``group`` on a free-text annotation is one file per distinct value). The rest
sit in between: ``group`` shards per distinct value unless a numeric
``threshold`` turns it into the bounded two-way ``under``/``atleast`` split the
ready-made pipelines rely on, ``sort`` shards only when max_count is given, and
``sample`` only when more than one sample size is requested.

A lane that has been fanned out is no longer a single stream, so the pipeline
**ends there**: the parts are the run's output and every one of them is
downloadable. Continuing would mean re-running the whole tail of the pipeline
once per part — unbounded work and disk for a request that never said how
many parts there would be — and every downstream structure here (the
PipelineStream's one-path-per-lane model, per-step pass/fail stats, the
retention funnel, the step-download endpoint) assumes one file per lane.

This module is the single place that answers "does this step, with these
parameters, split?" — the planner, the validator and the executor all ask it,
so the builder can never offer a step that the executor would then refuse.
"""

from __future__ import annotations

from typing import Any

#: Steps that always fan out, whatever their parameters.
ALWAYS_SPLITTING = frozenset({"SplitSeq.count"})

#: Steps that fan out only for certain parameter values.
CONDITIONAL_SPLITTING = frozenset(
    {"SplitSeq.group", "SplitSeq.sort", "SplitSeq.sample", "SplitSeq.samplepair"}
)

SPLITTING_STEPS = ALWAYS_SPLITTING | CONDITIONAL_SPLITTING


def _count_of(value: Any) -> int:
    """How many outputs a max_count parameter asks for."""
    if value is None:
        return 0
    if isinstance(value, (list, tuple, set)):
        return len([v for v in value if v is not None and v != ""])
    return 1


def step_splits_output(step_name: str, params: dict | None = None) -> bool:
    """True when this step writes more than one file per input lane."""
    if step_name in ALWAYS_SPLITTING:
        return True
    if step_name not in CONDITIONAL_SPLITTING:
        return False

    params = params or {}
    if step_name == "SplitSeq.group":
        # A numeric threshold is the bounded two-way "under"/"atleast" split
        # (the DUPCOUNT>=2 idiom the ready-made pipelines use, which carries
        # "atleast" forward). Without one it is a file per distinct value.
        return params.get("threshold") in (None, "")
    if step_name == "SplitSeq.sort":
        # sortSeqFile only starts a new part when max_count is supplied.
        return params.get("max_count") not in (None, "")
    # sample / samplepair write one file (or one pair) per requested size.
    return _count_of(params.get("max_count")) > 1


def split_reason(step_name: str, params: dict | None = None) -> str | None:
    """Human explanation of why a step ends the pipeline, or None."""
    if not step_splits_output(step_name, params):
        return None
    params = params or {}
    if step_name == "SplitSeq.count":
        max_count = params.get("max_count")
        detail = f" of at most {max_count} reads each" if max_count else ""
        return f"SplitSeq.count writes one file per partition{detail}"
    if step_name == "SplitSeq.group":
        field = params.get("field")
        detail = f" of {str(field).upper()}" if field else ""
        return (
            f"SplitSeq.group without a threshold writes one file per distinct "
            f"value{detail}"
        )
    if step_name == "SplitSeq.sort":
        return (
            f"SplitSeq.sort with max_count={params.get('max_count')} writes one "
            "file per partition"
        )
    return (
        f"{step_name} writes one output per requested sample size "
        f"({_count_of(params.get('max_count'))} requested)"
    )


def terminal_message(step_name: str, params: dict | None = None) -> str:
    """The message shown when a user tries to build past a splitting step."""
    reason = split_reason(step_name, params) or f"{step_name} splits the output"
    return (
        f"{reason}, so it has to be the last step: the pipeline cannot "
        "continue past a split. Download the parts and start a new run from "
        "the one you want."
    )
